### PR TITLE
feat(form-workspaces-be/3): add create workspace logic 

### PR DESCRIPTION
**Pull request by hanstirtaputra**
Builds on #4247 

Part of #4039 

## Solution
Add logic for POST `api/v3/admin/workspaces`endpoint

**Breaking Changes** 
- [x] No - this PR is backwards compatible

### CODE REVIEW

File: src/app/models/__tests__/workspace.server.model.spec.ts [hunk 1]
```diff
@@ -192,5 +192,37 @@ describe('Workspace Model', () => {
...
+        const actual = await Workspace.getWorkspaces(mockUserId)
+
+        expect(actual).toEqual([])
+      })
+
+      it('should return array of workspaces belonging to user', async () => {
+        const mockUserId = FORM_ADMIN_USER._id
+        const actual = await Workspace.getWorkspaces(mockUserId)
+
+        expect(actual.length).toEqual(1)
```
Comment (karrui) [inline src/app/models/__tests__/workspace.server.model.spec.ts]: doesn't seem to be a full test here?

File: src/app/models/workspace.server.model.ts [hunk 2]
```diff
@@ -75,6 +76,23 @@ const compileWorkspaceModel = (db: Mongoose): IWorkspaceModel => {
...
 
+  WorkspaceSchema.virtual('count').get(function (this: IWorkspaceSchema) {
+    return this.formIds.length
+  })
+
+  WorkspaceSchema.statics.getWorkspaces = async function (
+    admin: IUserSchema['_id'],
+  ): Promise<WorkspaceDto[]> {
+    return this.find({ admin: admin }).sort('title').exec()
+  }
```
Comment (karrui) [inline src/app/models/workspace.server.model.ts]: comment: these are duplicate code from the form-workspaces-be/2 PR yea?

**Review by karrui — APPROVED** [review]
lgtm, but seems to have content already reviewed in #4247

File: src/app/models/__tests__/workspace.server.model.spec.ts [hunk 3]
```diff
@@ -192,5 +192,37 @@ describe('Workspace Model', () => {
...
+
+    describe('getWorkspaces', () => {
+      it('should return empty array when user has no workspaces', async () => {
+        const mockUserId = new ObjectId()
+        const actual = await Workspace.getWorkspaces(mockUserId)
+
+        expect(actual).toEqual([])
+      })
+
+      it('should return array of workspaces belonging to user', async () => {
```
Comment (mantariksh) [inline src/app/models/__tests__/workspace.server.model.spec.ts]: what about sorting of the workspaces?

File: src/app/models/__tests__/workspace.server.model.spec.ts [hunk 4]
```diff
@@ -192,5 +192,37 @@ describe('Workspace Model', () => {
         mongoose.Error.ValidationError,
       )
     })
+
+    describe('getWorkspaces', () => {
```
Comment (mantariksh) [inline src/app/models/__tests__/workspace.server.model.spec.ts:113]: think this ideally belongs in the previous PR but no worries, no need to rearrange now

File: src/app/modules/workspace/__tests__/workspace.routes.spec.ts [hunk 5]
```diff
@@ -75,14 +77,83 @@ describe('workspaces.routes', () => {
...
+
+    it('should return 401 when user is not logged in', async () => {
+      await logoutSession(request)
+      const response = await request.post(CREATE_WORKSPACE_ENDPOINT)
+
+      expect(response.status).toEqual(401)
+      expect(response.body).toEqual({ message: 'User is unauthorized.' })
+    })
+
+    it('should return 500 when database errors occur', async () => {
```
Comment (mantariksh) [inline src/app/modules/workspace/__tests__/workspace.routes.spec.ts:164]: think we can probably separate this into the various database error cases, e.g. 409 for conflict error, 422 for validation error

File: src/app/models/workspace.server.model.ts (hunk 2 above)
Comment (hanstirtaputra) [reply src/app/models/workspace.server.model.ts]: Yep, it wasn't rebased yet when you saw it

File: src/app/models/__tests__/workspace.server.model.spec.ts (hunk 3 above)
Comment (hanstirtaputra) [reply src/app/models/__tests__/workspace.server.model.spec.ts]: Added test for sorting of workspaces

File: src/app/modules/workspace/__tests__/workspace.routes.spec.ts (hunk 5 above)
Comment (hanstirtaputra) [reply src/app/modules/workspace/__tests__/workspace.routes.spec.ts:164]: I've split it to the various error cases, and tested in the controller

File: src/app/models/__tests__/workspace.server.model.spec.ts (hunk 1 above)
Comment (hanstirtaputra) [reply src/app/models/__tests__/workspace.server.model.spec.ts]: Updated the test to expect for match instead!

**Review by mantariksh — APPROVED** [review]
merge conflicts but changes lgtm

**Pull request merged**